### PR TITLE
implementing a qual_name property for nodes

### LINKED ISSUE #7365: Add 'qual_name' property to Nodes
As per https://github.com/pytest-dev/pytest/pull/7265#issuecomment-637590707.

### PR DESCRIPTION
**Pull request by symonk**
initial attempt at adding a `qual_name` property to instances of `Node`.  This still needs some work and advice, namely these:

1. should this be '::' joining, or '.' ?
2. `listchain()` seems to be returning an `()` for functions inside classes, not sure (yet) if that is a bug or intentional behaviour and how to address that.

Still some work to do here, but would appreciate someone with more indepth node/collection to give their thoughts.

For some reason:

`<class 'test_node_qual_name_class_func.TestQual'>` has a `name` attr of `()`, not sure why that is?

closes #7365

### CODE REVIEW

File: src/_pytest/nodes.py [hunk 1]
```diff
@@ -172,6 +172,10 @@ def __init__(
         # own use. Currently only intended for internal plugins.
         self._store = Store()
 
+    @property
+    def qual_name(self) -> str:
+        return "::".join(node.name for node in self.listchain())
```
Comment (symonk) [inline src/_pytest/nodes.py:177]: `listnames` exists :) to get a list of names

Comment (nicoddemus) [reply src/_pytest/nodes.py:177]: A good we can use that then. 👍

**Review by nicoddemus — COMMENTED** [review]
> About `()`, we can just skip them, so this becomes:

The `()` is to denote an instance we can skip them.

> should this be '::' joining, or '.' ?

Ahh silly me of course, `::` is just to denote functions/classes, the rest of the node id uses `/`:

```
testing/acceptance_test.py::TestGeneralUsage::test_docstring_on_hookspec
```

So we need to figure out when to use `/` and when to use `::` (not clear at the moment for me, sorry, perhaps @RonnyPfannschmidt can chime in).

File: src/_pytest/nodes.py (hunk 1 above)
Comment (symonk) [reply src/_pytest/nodes.py:177]: Yep, ok to just drop `()` ? maybe we can inspect the parent type(s) to derive if it should be a `/` vs `::`

Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py]: At first glance this seems really strange

That kind of qualname is basically a potentially cut broken nodeid

**Comment by symonk** [comment]
ok @nicoddemus ignore that last commit, I was getting very confused as the suggestion on the ticket was not inline with what I think quar name is :)

**Comment by nicoddemus** [comment]
> ok @nicoddemus ignore that last commit, I was getting very confused as the suggestion on the ticket was not inline with what I think quar name is :)

Yeah sorry again about that!

**Comment by symonk** [comment]
> > ok @nicoddemus ignore that last commit, I was getting very confused as the suggestion on the ticket was not inline with what I think quar name is :)
> 
> Yeah sorry again about that!

no problem, it is my mistake.  P.s these tests are intentionally failing until I figure out what to do :)

File: src/_pytest/nodes.py (hunk 1 above)
Comment (symonk) [reply src/_pytest/nodes.py]: @RonnyPfannschmidt please see #7365, would appreciate your advice there in-line of recent comments

Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py]: I was under the impression that the qualname was supposed to be a python thing, if its going to be a cut up nodeid, it's going to be a problem

File: changelog/7365.feature.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Instances of :class:`Node <_pytest.nodes.Node>` now expose a `qual_name` property
```
Comment (bluetech) [inline changelog/7365.feature.rst:1]: The changelog should explain what this property is. Asking foremost for myself as a reviewer :grinning:

File: src/_pytest/nodes.py [hunk 3]
```diff
@@ -172,6 +172,10 @@ def __init__(
         # own use. Currently only intended for internal plugins.
         self._store = Store()
 
+    @property
+    def qual_name(self) -> str:
+        return "::".join([x for x in self.listnames() if x != "()"])
```
Comment (bluetech) [inline src/_pytest/nodes.py:177]:
```suggestion
        return "::".join(x for x in self.listnames() if x != "()")
```

**Comment by symonk** [comment]
need some follow up in #7365.  I'm unsure how to handle module names here, I (suspect) its best to just `replace(".py", "")`, unsure.  This is on hold for now

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus i think the concept of a qualname needs serious reconsideration to make sense for nodes

**Comment by nicoddemus** [comment]
> @nicoddemus i think the concept of a qualname needs serious reconsideration to make sense for nodes

Can you elaborate please?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus in python the `__qualname__` is a in module lookup identifier, the one prepared here is a incorrectly joined list of node names that even ignore node types

the in python way to determine qualname is already available on node.obj correctly, so i'm inclined to think we should not have a own version of it at all

**Comment by nicoddemus** [comment]
Agreed @RonnyPfannschmidt.

@symonk thanks for the work on this, but in the end it is better to just rely on the `node.obj`. 👍

**Pull request closed without merging**